kvgraph/graph.go

```diff
@@ -74,47 +57,91 @@ func (kgdb *KVInterfaceGDB) AddVertex(vertices []*gripql.Vertex) error {
 	return err
 }
 
+func insertVertex(tx kvi.KVBulkWrite, idx *kvindex.KVIndex, graph string, vertex *gripql.Vertex) error {
+	key := VertexKey(graph, vertex.Gid)
+	value, err := proto.Marshal(vertex)
+	if err != nil {
+		return nil
+	}
+	doc := map[string]interface{}{graph: vertexIdxStruct(vertex)}
+	if err := tx.Set(key, value); err != nil {
+		return fmt.Errorf("AddVertex Error %s", err)
+	}
+	if err := idx.AddDocTx(tx, vertex.Gid, doc); err != nil {
+		return fmt.Errorf("AddVertex Error %s", err)
+	}
+	return nil
+}
+
+func insertEdge(tx kvi.KVBulkWrite, idx *kvindex.KVIndex, graph string, edge *gripql.Edge) error {
+	eid := edge.Gid
+	var err error
+	var data []byte
+
+	data, err = proto.Marshal(edge)
+	if err != nil {
+		return err
+	}
+
+	src := edge.From
+	dst := edge.To
+	ekey := EdgeKey(graph, eid, src, dst, edge.Label, edgeSingle)
+	skey := SrcEdgeKey(graph, src, dst, eid, edge.Label, edgeSingle)
+	dkey := DstEdgeKey(graph, src, dst, eid, edge.Label, edgeSingle)
+
+	err = tx.Set(ekey, data)
+	if err != nil {
+		return err
+	}
+	err = tx.Set(skey, []byte{})
+	if err != nil {
+		return err
+	}
+	err = tx.Set(dkey, []byte{})
+	if err != nil {
+		return err
+	}
+	err = idx.AddDocTx(tx, eid, map[string]interface{}{graph: edgeIdxStruct(edge)})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
 // AddEdge adds an edge to the graph, if the id is not "" and in already exists
 // in the graph, it is replaced
 func (kgdb *KVInterfaceGDB) AddEdge(edges []*gripql.Edge) error {
 	err := kgdb.kvg.kv.BulkWrite(func(tx kvi.KVBulkWrite) error {
 		for _, edge := range edges {
-			eid := edge.Gid
-			var err error
-			var data []byte
-
-			data, err = proto.Marshal(edge)
-			if err != nil {
-				return err
-			}
-
-			src := edge.From
-			dst := edge.To
-			ekey := EdgeKey(kgdb.graph, eid, src, dst, edge.Label, edgeSingle)
-			skey := SrcEdgeKey(kgdb.graph, src, dst, eid, edge.Label, edgeSingle)
-			dkey := DstEdgeKey(kgdb.graph, src, dst, eid, edge.Label, edgeSingle)
+			insertEdge(tx, kgdb.kvg.idx, kgdb.graph, edge)
+		}
+		kgdb.kvg.ts.Touch(kgdb.graph)
+		return nil
+	})
+	return err
+}
 
-			err = tx.Set(ekey, data)
-			if err != nil {
-				return err
-			}
-			err = tx.Set(skey, []byte{})
-			if err != nil {
-				return err
-			}
-			err = tx.Set(dkey, []byte{})
-			if err != nil {
-				return err
+func (kgdb *KVInterfaceGDB) BulkAdd(stream <-chan *gripql.GraphElement) error {
```

Comment:
This method is missing the validate call on the vertices/edges. 

Can you add unit tests for BulkAdd, AddVertex, and AddEdge that attempts to load a invalid entity.